pkg/grafana/alertgroup-handler.go

```diff
@@ -212,18 +212,52 @@ func (h *AlertRuleGroupHandler) updateAlertRule(rule *models.ProvisionedAlertRul
 	return err
 }
 
-func (h *AlertRuleGroupHandler) putAlertRuleGroup(resource grizzly.Resource) error {
-	// TODO: Turn spec into a real models.AlertRuleGroup object
+func unMarshalAlertRuleGroup(resource grizzly.Resource) (*models.AlertRuleGroup, error) {
 	data, err := json.Marshal(resource.Spec())
 	if err != nil {
-		return err
+		return nil, err
 	}
 	var group models.AlertRuleGroup
 	err = json.Unmarshal(data, &group)
 	if err != nil {
-		return err
+		return nil, err
+	}
+	return &group, nil
+}
+
+func updateResourceToHaveUids(existing, resource grizzly.Resource) (*models.AlertRuleGroup, error) {
+	group, err := unMarshalAlertRuleGroup(existing)
+	if err != nil {
+		return nil, err
+	}
+	t := make(map[*string]string)
+	for _, rule := range group.Rules {
+		t[rule.Title] = rule.UID
+	}
+
+	group, err = unMarshalAlertRuleGroup(resource)
+	if err != nil {
+		return nil, err
+	}
+
+	uidsAdded := false
+	for _, rule := range group.Rules {
+		if uid, ok := t[rule.Title]; ok {
+			rule.UID = uid
+			uidsAdded = true
+		}
 	}
+	if !uidsAdded {
+		return nil, errors.New("No UIDs to add")
```

Comment:
This approach works for some usecases but fails when updating an alert rule group by adding a new rule. In this case, there is no existing rule to take the UID from, so `uidsAdded` will be false

I'd go for checking if every rule has a UID and fail if one does not

Reply:
Actually, how would that work?  This function is called from the PUT use-case, which does a GET by UID.  So if the UID doesn't exist, that GET fails.  Should we add a check before that GET also?

I'm not totally sure that this particular check is all that helpful any more -- we could probably just remove it.

Reply:
The reason why this fails is that we operate on a `AlertRuleGroup` level - not singular alert rules. This can result in the scenario where:

Remote:
* Rule A
Grizzly:
* Rule B

In this case, the rule applied from Grizzly will not have a UID added and the `No UIDs to add` error is returned.

I think removing the check is fine - what was your original intent when adding this check?

Reply:
I actually don't remember the original intention. This PR started out as a patch on a patch that added alert functionality before this handler existed.

I will remove this check, though I do think it would be worthwhile to check the existence of every rule for a UID before we call the GET.

Reply:
I have now added that second check.